### PR TITLE
Clustering

### PR DESCRIPTION
**Pull request by mommothazaz123**
### Summary
Adds the code necessary for running multiple clusters, including an IPC system using Redis Pub/Sub. Also moves a lot of env var accesses to a single config file, so it's easier to track all of those down.

Adds boto3 and aioredis as dependencies (for checking what ECS tasks are alive, and for async access of redis)

#### PR Type

- [ ] This PR is a code change that implements a feature request.
- [ ] This PR fixes an issue.
- [x] This PR adds a new feature that is not an open feature request.
- [ ] This PR is not a code change (e.g. documentation, README, ...)
#### Other
- [x] This PR is a breaking change (e.g. methods or parameters removed/renamed)
    - all redis database calls are now async
- [x] If code changes were made then they have been tested.
- [x] I have updated the documentation to reflect the changes.

### CODE REVIEW

**Comment by zmckinnon** [comment]
I added the python team in hopes that they can give some good feedback.

**Comment by JayCarney** [comment]
> in hopes that they can give some good feedback.

Man, no pressure or antyhing

File: cogs5e/models/dicecloud/client.py [hunk 1]
```diff
@@ -1,21 +1,19 @@
...
+from utils import config
 from .errors import InsertFailure, LoginFailure
 from .http import DicecloudHTTP
 
-TESTING = (os.environ.get("TESTING", False) or 'test' in sys.argv)
-NO_DICECLOUD = os.environ.get("NO_DICECLOUD", False)
-UNAME = os.getenv('DICECLOUD_USER', 'avrae') if not TESTING else credentials.test_dicecloud_user
-PWD = credentials.dicecloud_pass.encode() if not TESTING else credentials.test_dicecloud_pass.encode()
-API_KEY = credentials.dicecloud_token if not TESTING else credentials.test_dicecloud_token
+TESTING = config.TESTING
```
Comment (madcowfred) [inline cogs5e/models/dicecloud/client.py]: Bit of a code smell here, these are purely aliasing config.FOO to FOO - probably better to remove these and replace the various FOO use points with config.FOO?

File: utils/redisIO.py [hunk 2]
```diff
@@ -5,101 +5,190 @@
...
-        data = self.hget(key, field)
-        return json.loads(data) if data is not None else default
+    async def publish(self, channel, data):
+        return await self._db.publish(channel, data)
 
-    def jhset(self, key, field, value, **kwargs):
-        data = json.dumps(value, **kwargs)
-        return self.hset(key, field, data)
+
+class _PubSubMessageBase:
```
Comment (JayCarney) [inline utils/redisIO.py]: This should probably extend `abc.ABC` if it's intended to be abstract

File: utils/redisIO.py [hunk 3]
```diff
@@ -5,101 +5,190 @@
...
+
+    def to_dict(self):
+        inst = super(PubSubCommand, self).to_dict()
+        inst.update({"command": self.command, "args": self.args, "kwargs": self.kwargs})
+        return inst
+
+
+class PubSubReply(_PubSubMessageBase):
+    def __init__(self, id, sender, reply_to, data):
+        super(PubSubReply, self).__init__('reply', id, sender)
```
Comment (JayCarney) [inline utils/redisIO.py]: python 3 doesn't require passing in the class for super()
```suggestion
        super().__init__('reply', id, sender)
```

File: utils/redisIO.py [hunk 4]
```diff
@@ -5,101 +5,190 @@
...
+        self.reply_to = reply_to
+        self.data = data
+
+    @classmethod
+    def new(cls, bot, reply_to, data):
+        _id = str(uuid.uuid4())
+        return cls(_id, bot.cluster_id, reply_to, data)
+
+    def to_dict(self):
+        inst = super(PubSubReply, self).to_dict()
```
Comment (JayCarney) [inline utils/redisIO.py]:
```suggestion
        inst = super().to_dict()
```

File: utils/redisIO.py [hunk 5]
```diff
@@ -5,101 +5,190 @@
...
+    def to_dict(self):
+        return {"type": self.type, "id": self.id, "sender": self.sender}
+
+    def to_json(self):
+        return json.dumps(self.to_dict())
+
+
+class PubSubCommand(_PubSubMessageBase):
+    def __init__(self, id, sender, command, args, kwargs):
+        super(PubSubCommand, self).__init__('cmd', id, sender)
```
Comment (JayCarney) [inline utils/redisIO.py]:
```suggestion
        super().__init__('cmd', id, sender)
```

**Comment by JayCarney** [comment]
Random question, because I'm just missing some context; what leads to a ECS task not being alive and needing us to take over a dead cluster? just trying to understand the bigger picture.

**Comment by mommothazaz123** [comment]
@JayCarney For context overall, Discord requires bots that are in over 2500 guilds (servers) to connect via shards, then sends all activity from one server to a given shard based on ``SHARD_ID = (GUILD_ID >> 13) % NUM_SHARDS``. If an ECS task dies (which has happened if discord sends bad data, or randomly disconnects us and the reconnect logic fails), the takeover code ensures that a new task runs the same shards as the one that died, so you don't have two clients responding to the same events.

**Comment by JayCarney** [comment]
> @JayCarney For context overall, Discord requires bots that are in over 2500 guilds (servers) to connect via shards, then sends all activity from one server to a given shard based on `SHARD_ID = (GUILD_ID >> 13) % NUM_SHARDS`. If an ECS task dies (which has happened if discord sends bad data, or randomly disconnects us and the reconnect logic fails), the takeover code ensures that a new task runs the same shards as the one that died, so you don't have two clients responding to the same events.

Thanks for the context 👍

**Pull request merged**